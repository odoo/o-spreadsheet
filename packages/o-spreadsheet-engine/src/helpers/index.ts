export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
