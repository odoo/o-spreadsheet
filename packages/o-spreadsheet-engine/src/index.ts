export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
