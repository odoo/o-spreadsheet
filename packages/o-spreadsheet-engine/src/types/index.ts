export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
