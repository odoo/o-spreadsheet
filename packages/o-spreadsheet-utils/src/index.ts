export * from "./translation";
export * from "./uuid";
