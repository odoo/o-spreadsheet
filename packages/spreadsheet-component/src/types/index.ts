/**
 * State
 *
 * This file defines the basic types involved in maintaining the running state
 * of a o-spreadsheet.
 *
 * The most important exported values are:
 * - interface GridState: the internal type of the state managed by the model
 */

import { Chart } from "chart.js";

export * from "./autofill";
export * from "./cells";
export * from "./chart/chart";
export * from "./clipboard";
export * from "./collaborative/revisions";
export * from "./collaborative/session";
export * from "./commands";
export * from "./conditional_formatting";
export * from "./currency";
export * from "./data_validation";
export * from "./env";
export * from "./figure";
export * from "./format";
export * from "./functions";
export * from "./getters";
export * from "./history";
export * from "./locale";
export * from "./misc";
export * from "./pivot";
export * from "./pivot_runtime";
export * from "./range";
export * from "./rendering";
export * from "./table";
export * from "./workbook_data";

declare global {
  interface Window {
    Chart: typeof Chart;
  }
}
