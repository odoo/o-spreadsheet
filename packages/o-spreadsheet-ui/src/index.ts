import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}