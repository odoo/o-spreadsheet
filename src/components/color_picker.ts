import * as owl from "@odoo/owl";
import { SpreadsheetEnv } from "../types/env";
const Component = owl.Component;
const { css, xml } = owl.tags;

const COLORS = [
  [
    "#000000",
    "#434343",
    "#666666",
    "#999999",
    "#b7b7b7",
    "#cccccc",
    "#d9d9d9",
    "#efefef",
    "#f3f3f3",
    "#ffffff",
  ],
  [
    "#980000",
    "#ff0000",
    "#ff9900",
    "#ffff00",
    "#00ff00",
    "#00ffff",
    "#4a86e8",
    "#0000ff",
    "#9900ff",
    "#ff00ff",
  ],
  [
    "#e6b8af",
    "#f4cccc",
    "#fce5cd",
    "#fff2cc",
    "#d9ead3",
    "#d0e0e3",
    "#c9daf8",
    "#cfe2f3",
    "#d9d2e9",
    "#ead1dc",
  ],
  [
    "#dd7e6b",
    "#ea9999",
    "#f9cb9c",
    "#ffe599",
    "#b6d7a8",
    "#a2c4c9",
    "#a4c2f4",
    "#9fc5e8",
    "#b4a7d6",
    "#d5a6bd",
  ],
  [
    "#cc4125",
    "#e06666",
    "#f6b26b",
    "#ffd966",
    "#93c47d",
    "#76a5af",
    "#6d9eeb",
    "#6fa8dc",
    "#8e7cc3",
    "#c27ba0",
  ],
  [
    "#a61c00",
    "#cc0000",
    "#e69138",
    "#f1c232",
    "#6aa84f",
    "#45818e",
    "#3c78d8",
    "#3d85c6",
    "#674ea7",
    "#a64d79",
  ],
  [
    "#85200c",
    "#990000",
    "#b45f06",
    "#bf9000",
    "#38761d",
    "#134f5c",
    "#1155cc",
    "#0b5394",
    "#351c75",
    "#741b47",
  ],
  [
    "#5b0f00",
    "#660000",
    "#783f04",
    "#7f6000",
    "#274e13",
    "#0c343d",
    "#1c4587",
    "#073763",
    "#20124d",
    "#4c1130",
  ],
];

export class ColorPicker extends Component<any, SpreadsheetEnv> {
  static template = xml/* xml */ `
  <div class="o-color-picker" t-on-click="onColorClick">
    <div class="o-color-picker-line" t-foreach="COLORS" t-as="colors" t-key="colors">
      <t t-foreach="colors" t-as="color" t-key="color">
        <div class="o-color-picker-line-item" t-att-data-color="color" t-attf-style="background-color:{{color}};"></div>
      </t>
    </div>
  </div>`;

  static style = css/* scss */ `
    .o-color-picker {
      position: absolute;
      top: calc(100% + 5px);
      left: 0;
      z-index: 10;
      box-shadow: 1px 2px 5px 2px rgba(51, 51, 51, 0.15);
      background-color: white;
      padding: 6px 0px;

      .o-color-picker-line {
        display: flex;
        padding: 3px 6px;
        .o-color-picker-line-item {
          width: 18px;
          height: 18px;
          margin: 0px 2px;
          border-radius: 50px;
          border: 1px solid #c0c0c0;
          &:hover {
            background-color: rgba(0, 0, 0, 0.08);
            outline: 1px solid gray;
          }
        }
      }
    }
  `;
  COLORS = COLORS;

  onColorClick(ev: MouseEvent) {
    const color = (ev.target as HTMLElement).dataset.color;
    if (color) {
      this.trigger("color-picked", { color });
    }
  }
}
