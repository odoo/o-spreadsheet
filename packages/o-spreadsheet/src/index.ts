import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };