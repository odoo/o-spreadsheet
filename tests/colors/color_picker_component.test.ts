import { Model } from "../../src";
import { ColorPicker, ColorPickerProps } from "../../src/components/color_picker/color_picker";
import { toHex } from "../../src/helpers";
import { Color } from "../../src/types";
import { setStyle } from "../test_helpers/commands_helpers";
import {
  getElComputedStyle,
  setInputValueAndTrigger,
  simulateClick,
} from "../test_helpers/dom_helper";
import { mountComponentWithPortalTarget } from "../test_helpers/helpers";
import { mockGetBoundingClientRect } from "../test_helpers/mock_helpers";

mockGetBoundingClientRect({
  "o-spreadsheet": () => ({ x: 0, y: 0, width: 1000, height: 1000 }),
});

let fixture: HTMLElement;

async function mountColorPicker(partialProps: Partial<ColorPickerProps> = {}, model = new Model()) {
  const props = {
    onColorPicked: partialProps.onColorPicked || (() => {}),
    currentColor: partialProps.currentColor || "#000000",
    maxHeight: partialProps.maxHeight !== undefined ? partialProps.maxHeight : 1000,
    anchorRect: partialProps.anchorRect || { x: 0, y: 0, width: 0, height: 0 },
  };
  ({ fixture } = await mountComponentWithPortalTarget(ColorPicker, { model, props }));
}

test("Color picker is correctly positioned", async () => {
  await mountColorPicker({ anchorRect: { x: 100, y: 100, width: 50, height: 50 } });
  expect(getElComputedStyle(".o-popover", "left")).toEqual("100px");
  expect(getElComputedStyle(".o-popover", "top")).toEqual("150px");
});

describe("Color Picker buttons", () => {
  test("Click on 'reset' button reset the cell style", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ currentColor: "#45818e", onColorPicked });
    await simulateClick(".o-cancel");
    expect(onColorPicked).toHaveBeenCalledWith("");
  });

  test("Clicking on '+', custom section except custom added colors must toggle the custom color picker", async () => {
    await mountColorPicker();
    const elements = fixture.querySelectorAll(".o-color-picker-toggler");
    for (const element of elements) {
      await simulateClick(element);
      expect(fixture.querySelector(".o-custom-selector")).toBeDefined();
      await simulateClick(element);
      expect(fixture.querySelector(".o-custom-selector")).toBeNull();
    }
  });

  test("Full component rendering", async () => {
    await mountColorPicker();
    await simulateClick(".o-color-picker-toggler-sign");
    expect(fixture.querySelector(".o-color-picker")).toMatchSnapshot();
  });

  test("Can pick a standard color", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ onColorPicked });
    await simulateClick("div[data-color='#FF9900']");
    expect(onColorPicked).toHaveBeenCalledWith("#FF9900");
  });

  test("Can pick a custom color in the gradient", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ onColorPicked });
    await simulateClick(".o-color-picker-toggler");
    await simulateClick(".o-gradient");
    const inputCodeEl = fixture.querySelector(".o-custom-input-preview input") as HTMLInputElement;
    const previewColor = toHex(getElComputedStyle(".o-color-preview", "backgroundColor"));
    const inputColorCode = inputCodeEl.value;
    expect(previewColor).toBeSameColorAs(inputColorCode);
    await simulateClick(".o-add-button");
    expect(onColorPicked).toHaveBeenCalledWith(inputColorCode);
  });

  test("Can choose a custom color with the input", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ onColorPicked });
    await simulateClick(".o-color-picker-toggler");
    await simulateClick(".o-gradient");
    const color = "#12EF78";
    await setInputValueAndTrigger(".o-custom-input-preview input", color);
    const previewColor = toHex(getElComputedStyle(".o-color-preview", "backgroundColor"));
    // hex <-> hsla is not a bijection, this specific color
    // is not exactly the same when processed
    expect(previewColor).toBeSameColorAs(color, 0.01);
    await simulateClick(".o-add-button");
    expect(onColorPicked).toHaveBeenCalledWith(color);
  });

  test("Color from the input is sanitized", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ onColorPicked });
    await simulateClick(".o-color-picker-toggler");
    await simulateClick(".o-gradient");
    const color = "12ef78";
    await setInputValueAndTrigger(".o-custom-input-preview input", color);
    await simulateClick(".o-add-button");
    expect(onColorPicked).toHaveBeenCalledWith(toHex(color));
  });

  test("Cannot input an invalid color code", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ onColorPicked });
    await simulateClick(".o-color-picker-toggler");
    const target = document.querySelector(".o-custom-input-preview input");
    await setInputValueAndTrigger(target, "this is not a color");

    expect(fixture.querySelector(".o-wrong-color")).not.toBeNull();
    const addButton = fixture.querySelector(".o-add-button")!;
    expect(addButton.classList).toContain("o-disabled");
    await simulateClick(addButton);
    expect(onColorPicked).not.toHaveBeenCalled();
  });

  test("warning border of invalid color disappears when selecting on the gradient", async () => {
    const onColorPicked = jest.fn();
    await mountColorPicker({ onColorPicked });
    await simulateClick(".o-color-picker-toggler");
    await setInputValueAndTrigger(".o-custom-input-preview input", "this is not a color");
    await simulateClick(".o-add-button");
    expect(fixture.querySelector(".o-wrong-color")).not.toBeNull();
    await simulateClick(".o-gradient");
    expect(fixture.querySelector(".o-wrong-color")).toBeNull();
    await simulateClick(".o-add-button");
    expect(onColorPicked).toHaveBeenCalled();
  });

  test("initial standard color", async () => {
    await mountColorPicker({ currentColor: "#45818e" });
    const color = fixture.querySelector("div[data-color='#45818E']") as HTMLElement;
    expect(color?.textContent).toBe(" ✓ ");
  });

  test("initial custom color", async () => {
    const model = new Model();
    setStyle(model, "A1", { fillColor: "#123456" });
    await mountColorPicker({ currentColor: "#123456" }, model);
    const color = fixture.querySelector("div[data-color='#123456']") as HTMLElement;
    expect(color?.textContent).toBe(" ✓ ");
  });

  test("wrong initial color", async () => {
    await mountColorPicker({ currentColor: "azeazeaze" });
    await simulateClick(".o-color-picker-toggler");
    const inputCodeEl = fixture.querySelector(".o-custom-input-preview input") as HTMLInputElement;
    expect(inputCodeEl.value).toBe("");
  });

  test("color picker disappears when maxHeight is 0", async () => {
    await mountColorPicker({ currentColor: "#45818e", maxHeight: 0 });
    const picker = fixture.querySelector<HTMLElement>(".o-color-picker")!;
    expect(picker.style["display"]).toEqual("none");
  });

  test.each([
    "#fff",
    "fff",
    "#FFFFFF00", // Hex + alpha
  ])("Can input a custom HEX code, alpha is ignored", async (hexCode) => {
    await mountColorPicker();
    await simulateClick(".o-color-picker-toggler");

    const inputTarget = fixture.querySelector(".o-custom-input-preview input")!;
    await setInputValueAndTrigger(inputTarget, hexCode as Color);
    expect((inputTarget as HTMLInputElement).value).toBeSameColorAs(hexCode.slice(0, 7));
    const addButton = fixture.querySelector(".o-add-button")!;
    expect(addButton.classList).not.toContain("o-disabled");
  });

  test.each([
    "rgb(1,1,1)", // rgb
    "rgb(1,1,1,0.5)", // rgba
  ])("refuse non strictly HEX codes", async (hexCode) => {
    await mountColorPicker();
    await simulateClick(".o-color-picker-toggler");

    const inputTarget = fixture.querySelector(".o-custom-input-preview input")!;
    await setInputValueAndTrigger(inputTarget, hexCode as Color);
    expect((inputTarget as HTMLInputElement).value).toBe(hexCode.slice(0, 7));
    const addButton = fixture.querySelector(".o-add-button")!;
    expect(addButton.classList).toContain("o-disabled");
  });
});
