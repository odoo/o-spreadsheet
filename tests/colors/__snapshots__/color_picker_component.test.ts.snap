// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color Picker buttons Full component rendering 1`] = `
<div
  class="o-color-picker"
  style=""
>
  <div
    class="o-color-picker-section-name"
  >
    Standard
  </div>
  <div
    class="colors-grid"
  >
    <div
      class="o-color-picker-line-item"
      data-color="#000000"
      style="background-color:#000000;"
    >
      <div
        align="center"
        style="color:#FFFFFF"
      >
         âœ“ 
      </div>
    </div>
    <div
      class="o-color-picker-line-item"
      data-color="#434343"
      style="background-color:#434343;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#666666"
      style="background-color:#666666;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#999999"
      style="background-color:#999999;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#B7B7B7"
      style="background-color:#B7B7B7;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#CCCCCC"
      style="background-color:#CCCCCC;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#D9D9D9"
      style="background-color:#D9D9D9;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#EFEFEF"
      style="background-color:#EFEFEF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#F3F3F3"
      style="background-color:#F3F3F3;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FFFFFF"
      style="background-color:#FFFFFF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#980000"
      style="background-color:#980000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FF0000"
      style="background-color:#FF0000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FF9900"
      style="background-color:#FF9900;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FFFF00"
      style="background-color:#FFFF00;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#00FF00"
      style="background-color:#00FF00;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#00FFFF"
      style="background-color:#00FFFF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#4A86E8"
      style="background-color:#4A86E8;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#0000FF"
      style="background-color:#0000FF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#9900FF"
      style="background-color:#9900FF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FF00FF"
      style="background-color:#FF00FF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#E6B8AF"
      style="background-color:#E6B8AF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#F4CCCC"
      style="background-color:#F4CCCC;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FCE5CD"
      style="background-color:#FCE5CD;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FFF2CC"
      style="background-color:#FFF2CC;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#D9EAD3"
      style="background-color:#D9EAD3;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#D0E0E3"
      style="background-color:#D0E0E3;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#C9DAF8"
      style="background-color:#C9DAF8;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#CFE2F3"
      style="background-color:#CFE2F3;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#D9D2E9"
      style="background-color:#D9D2E9;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#EAD1DC"
      style="background-color:#EAD1DC;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#DD7E6B"
      style="background-color:#DD7E6B;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#EA9999"
      style="background-color:#EA9999;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#F9CB9C"
      style="background-color:#F9CB9C;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FFE599"
      style="background-color:#FFE599;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#B6D7A8"
      style="background-color:#B6D7A8;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#A2C4C9"
      style="background-color:#A2C4C9;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#A4C2F4"
      style="background-color:#A4C2F4;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#9FC5E8"
      style="background-color:#9FC5E8;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#B4A7D6"
      style="background-color:#B4A7D6;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#D5A6BD"
      style="background-color:#D5A6BD;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#CC4125"
      style="background-color:#CC4125;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#E06666"
      style="background-color:#E06666;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#F6B26B"
      style="background-color:#F6B26B;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#FFD966"
      style="background-color:#FFD966;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#93C47D"
      style="background-color:#93C47D;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#76A5AF"
      style="background-color:#76A5AF;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#6D9EEB"
      style="background-color:#6D9EEB;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#6FA8DC"
      style="background-color:#6FA8DC;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#8E7CC3"
      style="background-color:#8E7CC3;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#C27BA0"
      style="background-color:#C27BA0;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#A61C00"
      style="background-color:#A61C00;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#CC0000"
      style="background-color:#CC0000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#E69138"
      style="background-color:#E69138;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#F1C232"
      style="background-color:#F1C232;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#6AA84F"
      style="background-color:#6AA84F;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#45818E"
      style="background-color:#45818E;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#3C78D8"
      style="background-color:#3C78D8;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#3D85C6"
      style="background-color:#3D85C6;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#674EA7"
      style="background-color:#674EA7;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#A64D79"
      style="background-color:#A64D79;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#85200C"
      style="background-color:#85200C;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#990000"
      style="background-color:#990000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#B45F06"
      style="background-color:#B45F06;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#BF9000"
      style="background-color:#BF9000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#38761D"
      style="background-color:#38761D;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#134F5C"
      style="background-color:#134F5C;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#1155CC"
      style="background-color:#1155CC;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#0B5394"
      style="background-color:#0B5394;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#351C75"
      style="background-color:#351C75;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#741B47"
      style="background-color:#741B47;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#5B0F00"
      style="background-color:#5B0F00;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#660000"
      style="background-color:#660000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#783F04"
      style="background-color:#783F04;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#7F6000"
      style="background-color:#7F6000;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#274E13"
      style="background-color:#274E13;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#0C343D"
      style="background-color:#0C343D;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#1C4587"
      style="background-color:#1C4587;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#073763"
      style="background-color:#073763;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#20124D"
      style="background-color:#20124D;"
    />
    <div
      class="o-color-picker-line-item"
      data-color="#4C1130"
      style="background-color:#4C1130;"
    />
    
  </div>
  <div
    class="o-separator"
  />
  <div
    class="o-color-picker-section-name o-color-picker-toggler"
  >
    <span>
      Custom
    </span>
  </div>
  <div
    class="colors-grid o-color-picker-toggler"
  >
    <div
      class="o-color-picker-line-item o-color-picker-toggler-button"
    >
      <div
        class="o-color-picker-toggler-sign"
      >
        <svg
          class="o-icon plus"
          viewBox="0 0 18 18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0h2v8h8v2h-8v8H8v-8H0V8h8"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    
    
  </div>
  <div
    class="o-custom-selector"
  >
    <div
      class="o-gradient"
      style="background:hsl(0 100% 50%); "
    >
      <div
        class="saturation w-100 h-100 position-absolute pe-none"
      />
      <div
        class="lightness w-100 h-100 position-absolute pe-none"
      />
      <div
        class="magnifier pe-none"
        style="left:-8px; top:178px; background:#000000; "
      />
    </div>
    <div
      class="o-hue-container"
    >
      <div
        class="o-hue-picker"
      />
      <div
        class="o-hue-slider pe-none"
        style="margin-left:-8px; "
      >
        <div
          class="o-icon fa-small"
        >
          <i
            class="fa fa-caret-up"
          />
        </div>
      </div>
    </div>
    <div
      class="o-custom-input-preview"
    >
      <input
        maxlength="7"
        type="text"
      />
      <div
        class="o-color-preview"
        style="background-color:#000000; "
      />
    </div>
    <div
      class="o-custom-input-buttons"
    >
      <button
        class="o-add-button"
      >
         Add 
      </button>
    </div>
  </div>
  
  <div
    class="o-separator"
  />
  <div
    class="o-buttons"
  >
    <button
      class="o-cancel"
    >
      No Color
    </button>
  </div>
</div>
`;
